java/ql/src/semmle/code/java/dataflow/internal/DataFlowImplCommon.qll

```diff
@@ -3,481 +3,685 @@ import DataFlowImplSpecific::Public
-    )
-  }
-
-  /**
-   * Holds if `p` can flow to `node` in the same callable using only
-   * value-preserving steps, not taking call contexts into account.
-   */
-  private predicate parameterValueFlowCand(ParameterNode p, Node node) {
-    p = node
-    or
-    exists(Node mid |
-      parameterValueFlowCand(p, mid) and
-      simpleLocalFlowStep(mid, node) and
-      compatibleTypes(p.getType(), node.getType())
-    )
-    or
-    // flow through a callable
-    exists(Node arg |
-      parameterValueFlowCand(p, arg) and
-      argumentValueFlowsThroughCand(arg, node) and
-      compatibleTypes(p.getType(), node.getType())
-    )
-  }
-
-  /**
-   * Holds if `p` can flow to a return node of kind `kind` in the same
-   * callable using only value-preserving steps, not taking call contexts
-   * into account.
-   */
-  private predicate parameterValueFlowsThroughCand(ParameterNode p, ReturnKind kind) {
-    parameterValueFlowCand(p, getAReturnNodeOfKind(kind))
-  }
-
-  pragma[nomagic]
-  private predicate argumentValueFlowsThroughCand0(
-    DataFlowCall call, ArgumentNode arg, ReturnKind kind
-  ) {
-    exists(ParameterNode param | viableParamArg(call, param, arg) |
-      parameterValueFlowsThroughCand(param, kind)
-    )
-  }
-
-  /**
-   * Holds if `arg` flows to `out` through a call using only value-preserving steps,
-   * not taking call contexts into account.
-   */
-  private predicate argumentValueFlowsThroughCand(ArgumentNode arg, OutNode out) {
-    exists(DataFlowCall call, ReturnKind kind | argumentValueFlowsThroughCand0(call, arg, kind) |
-      out = getAnOutNode(call, kind) and
-      compatibleTypes(arg.getType(), out.getType())
-    )
-  }
-
-  /**
-   * Holds if `arg` is the `i`th argument of `call` inside the callable
-   * `enclosing`, and `arg` may flow through `call`.
-   */
-  pragma[noinline]
-  private predicate argumentOf(
-    DataFlowCall call, int i, ArgumentNode arg, DataFlowCallable enclosing
-  ) {
-    arg.argumentOf(call, i) and
-    argumentValueFlowsThroughCand(arg, _) and
-    enclosing = arg.getEnclosingCallable()
-  }
+  private module Cached {
+    /**
+     * Holds if `p` is the `i`th parameter of a viable dispatch target of `call`.
+     * The instance parameter is considered to have index `-1`.
+     */
+    pragma[nomagic]
+    private predicate viableParam(DataFlowCall call, int i, ParameterNode p) {
+      p.isParameterOf(viableCallable(call), i)
+    }
+
+    /**
+     * Holds if `arg` is a possible argument to `p` in `call`, taking virtual
+     * dispatch into account.
+     */
+    cached
+    predicate viableParamArg(DataFlowCall call, ParameterNode p, ArgumentNode arg) {
+      exists(int i |
+        viableParam(call, i, p) and
+        arg.argumentOf(call, i)
+      )
+    }
 
-  pragma[noinline]
-  private ParameterNode getAParameter(DataFlowCallable c) { result.getEnclosingCallable() = c }
+    private module FlowThrough_v1 {
+      private predicate step = simpleLocalFlowStep/2;
```

Comment:
This alias is only called twice, do we really need it?

Reply:
The `FlowThrough_*` modules are essentially two instantiations of an implicit parameterized module.  The `step` relation is the input, and the only thing that differs between `_v1` and `_v2`.

Reply:
That's also what I reckoned, but I think we should then at least have a comment saying that, when not using pyrameterization.